lib/connection/pool.js

```diff
@@ -635,34 +635,53 @@ Pool.prototype.unref = function() {
 var events = ['error', 'close', 'timeout', 'parseError', 'connect', 'message'];
 
 // Destroy the connections
-function destroy(self, connections) {
+function destroy(self, connections, options, callback) {
+  let connectionCount = connections.length;
+  function connectionDestroyed() {
+    connectionCount--;
+    if (connectionCount > 0) {
+      return;
+    }
+
+    // Zero out all connections
+    self.inUseConnections = [];
+    self.availableConnections = [];
+    self.connectingConnections = 0;
+
+    // Set state to destroyed
+    stateTransition(self, DESTROYED);
+    if (typeof callback === 'function') {
+      callback(null, null);
+    }
+  }
+
+  if (connectionCount === 0) {
+    connectionDestroyed();
+    return;
+  }
+
   // Destroy all connections
-  connections.forEach(function(c) {
-    // Remove all listeners
+  connections.forEach(conn => {
     for (var i = 0; i < events.length; i++) {
-      c.removeAllListeners(events[i]);
+      conn.removeAllListeners(events[i]);
     }
-    // Destroy connection
-    c.destroy();
-  });
 
-  // Zero out all connections
-  self.inUseConnections = [];
-  self.availableConnections = [];
-  self.connectingConnections = 0;
-
-  // Set state to destroyed
-  stateTransition(self, DESTROYED);
+    conn.destroy(options, connectionDestroyed);
```

Comment:
Is it possible for `pool.destroy()` to be called synchronously anymore?

Reply:
I think the legacy behavior is maintained in the event a callback is not provided. The only behavioral change there is that previously the cleanup would happen most likely before the individual connections were cleaned up (though they would be prevented from being garbage collected because of their references here). Are you worried about something in particular?

Reply:
I was mostly concerned about the legacy behavior.